File: apps/authentication/src/functions/AuthHandler.ts
'use strict';

import {Handler} from 'aws-lambda';
import {RegisterRequest} from '../model/request/RegisterRequest';
import {LoginRequest} from '../model/request/LoginRequest';
import {LogoutRequest} from '../model/request/LogoutRequest';

import {AuthenticationService} from '../service/AuthenticationService';
import {ConfirmSignupRequest} from "../model/request/ConfirmSignupRequest";
import {SendCodeRequest} from "../model/request/SendCodeRequest";
import {ConfirmCodeRequest} from "../model/request/ConfirmCodeRequest";
import {UserService} from "../service/UserService";
import {UpdateRequest} from "../model/request/UpdateRequest";
import {ForgotPasswordRequest} from "../model/request/ForgotPasswordRequest";
import {ResetPasswordRequest} from "../model/request/ResetPasswordRequest";

const authenticationService = new AuthenticationService()
const userService = new UserService()

export const getMinAppVersions: Handler = async (event: any) => {
    try {
        console.log("Received request to get min app versions")
        return createResponse(200, {
            "ios": process.env.MIN_IOS_VERSION
        })
    } catch (e: any) {
        return createResponse(e.statusCode, e.message)
    }
}

export const register: Handler = async (event: any) => {
    try {
        console.log("Received request to create an account")
        const body = typeof event.body === 'object' ? event.body : JSON.parse(event.body)
        const request = new RegisterRequest(body.username, body.password)
        request.validate()
        const registerResponse = await authenticationService.register(request)
        return createResponse(200, registerResponse)
    } catch (e: any) {
        console.log(`Failed to register account for ${JSON.parse(event.body)} due to ${e.message}`)
        return createResponse(e.statusCode, e.message)
    }
}

export const confirm: Handler = async (event: any) => {
    try {
        console.log("Received request to confirm an account")
        const body = typeof event.body === 'object' ? event.body : JSON.parse(event.body)
        const request = new ConfirmSignupRequest(body.username, body.confirmationCode)
        request.validate()
        const confirmResponse = await authenticationService.confirm(request)

        // Once user account has been confirmed we can create the user record in the database
        // Need to do this before returning because user will be logging in immediately after
        // this and requesting data
        console.log("User confirmed.. calling user service to create record", body.username)
        const user = await authenticationService.getUserByPhone(body.username)
        console.log("User in cognito", user)
        if (!user) {
            console.log("Didn't find user with expected phone number")
            return createResponse(500, {})
        }
        await userService.createUser(user?.userId, user?.phoneNumber)

        return createResponse(200, confirmResponse)
    } catch (e: any) {
        console.log(`Failed to confirm account for ${JSON.parse(event.body)} due to ${e.message}`)
        return createResponse(e.statusCode, e.message)
    }
}

export const login: Handler = async (event: any) => {
    try {
        console.log("Received request to login")
        const body = typeof event.body === 'object' ? event.body : JSON.parse(event.body)
        const request = new LoginRequest(body.username, body.password)
        request.validate()
        const loginResponse = await authenticationService.login(request)
        return createResponse(200, loginResponse)
    } catch (e: any) {
        console.log(`Failed to login due to ${e.message}`)
        return createResponse(e.statusCode, e.message)
    }
}

export const logout: Handler = async (event: any) => {
    try {
        console.log("Received request to logout")
        const accessToken = event.headers.AccessToken || event.headers.accesstoken
        const request = new LogoutRequest(accessToken)
        request.validate()
        const logoutResponse = await authenticationService.logout(request)
        return createResponse(200, logoutResponse)
    } catch (e: any) {
        return createResponse(e.statusCode, e.message)
    }
}

export const forgotPassword: Handler = async (event: any) => {
    try {
        console.log("Received request to forgot password")
        const body = typeof event.body === 'object' ? event.body : JSON.parse(event.body)
        const request = new ForgotPasswordRequest(body.username)
        request.validate()
        await authenticationService.forgotPassword(request)
        return createResponse(200, {
            "message": "A verification code has been sent to that number if it exists."
        })
    } catch (e: any) {
        return createResponse(e.statusCode, e.message)
    }
}

export const confirmForgotPassword: Handler = async (event: any) => {
    try {
        console.log("Received request to confirm a forgotten password")
        const body = typeof event.body === 'object' ? event.body : JSON.parse(event.body)
        const request = new ResetPasswordRequest(body.username, body.code, body.password)
        request.validate()
        await authenticationService.confirmForgotPassword(request)
        return createResponse(200, {
            "message": "Password has been reset"
        })
    } catch (e: any) {
        console.log(`Failed to confirm password due to ${e.message}`)
        return createResponse(e.statusCode, e.message)
    }
}


export const sendCode: Handler = async (event: any) => {
    try {
        console.log("Received request to send a new verification code")
        const {attribute} = (event.queryStringParameters || {})
        const accessToken = event.headers.AccessToken || event.headers.accesstoken
        const request = new SendCodeRequest(accessToken, attribute)
        request.validate()
        const response = await authenticationService.sendCode(request)
        return createResponse(200, response)
    } catch (e: any) {
        return createResponse(e.statusCode, e.message)
    }
}

export const confirmCode: Handler = async (event: any) => {
    try {
        console.log("Received request to confirm a verification code")
        const {attribute, code} = (event.queryStringParameters || {})
        const accessToken = event.headers.AccessToken || event.headers.accesstoken
        const request = new ConfirmCodeRequest(accessToken, attribute, code)
        request.validate()

        // If we successfully validate the email we need to update the user
        const userId = await authenticationService.confirmCode(request)
        await userService.confirmUserEmail(userId)

        return createResponse(200, {"message": "Success"})
    } catch (e: any) {
        console.log(`Failed to confirm code for ${event.queryStringParameters} due to ${e.message}`)
        return createResponse(e.statusCode, e.message)
    }
}

export const updateUser: Handler = async (event) => {
    try{
        console.log("Received request to update user")

        // validate incoming request is good
        const body = typeof event.body === 'object' ? event.body : JSON.parse(event.body)
        const request = new UpdateRequest(body.email)
        request.validate()

        const {userId} = event.pathParameters
        const updatedUser = await authenticationService.updateUser(userId, request)

        return createResponse(200, updatedUser)
    } catch (e: any) {
        return createResponse(e.statusCode, e.message)
    }
}

function createResponse(statusCode: number, body: object) {
    return {
        statusCode: statusCode || 500,
        body: JSON.stringify(body)
    }
}

File: apps/authentication/src/functions/auth_handler.py
import json
import logging
import os

from src.functions.handler_utils import create_response, exception_handler
from src.model.requests import LoginRequest, RegisterRequest, ConfirmSignupRequest, RefreshLoginRequest, LogoutRequest, \
    ResendSignupCodeRequest, ForgotPasswordRequest, ResetPasswordRequest, SendCodeRequest, ConfirmCodeRequest, \
    UpdateRequest
from src.service.authentication_service import AuthenticationService
from src.service.user_service import UserService

logging.basicConfig(level=logging.INFO)
logging.getLogger().setLevel(level=logging.INFO)
logger = logging.getLogger(__name__)

auth_service = AuthenticationService()
user_service = UserService()


@exception_handler
def get_min_app_versions(event, context):
    return create_response(status_code=200, body={
        "ios": os.environ["MIN_IOS_VERSION"]
    })


@exception_handler
def register(event, context):
    body = json.loads(event.get('body', '{}'))
    request = RegisterRequest(username=body.get("username"), password=body.get("password"))
    response = auth_service.register(request=request)
    logger.info(f"Created new user: {request.username}")
    return create_response(200, response)


@exception_handler
def resend_signup_code(event, context):
    body = json.loads(event.get('body', '{}'))
    request = ResendSignupCodeRequest(username=body.get("username"))
    auth_service.resend_signup_code(request=request)
    return create_response(200, {})


@exception_handler
def login(event, context):
    body = json.loads(event.get('body', '{}'))
    request = LoginRequest(username=body.get("username"), password=body.get("password"))
    response = auth_service.login(request=request)
    logger.info(f"Logged in user: {request.username}")
    return create_response(200, response)


@exception_handler
def refresh_login(event, context):
    headers = event.get('headers', {})
    refresh_token = headers.get("RefreshToken") or headers.get("refreshtoken")
    request = RefreshLoginRequest(refresh_token=refresh_token)
    response = auth_service.refresh_login(request=request)
    return create_response(200, response)


@exception_handler
def logout(event, context):
    headers = event.get('headers', {})
    access_token = headers.get("AccessToken") or headers.get("accesstoken")
    request = LogoutRequest(access_token=access_token)
    auth_service.logout(request=request)
    return create_response(200, {})


@exception_handler
def confirm(event, context):
    body = json.loads(event.get('body', '{}'))
    request = ConfirmSignupRequest(username=body.get("username"), confirmation_code=body.get("confirmationCode"))
    auth_service.confirm(request=request)

    # We need the user id to create the user profile, go back to Cognito to get it
    logger.info(f"User {request.username} confirmed.. calling user auth_service to create record")
    user = auth_service.get_user_by_phone(phone_number=request.username)

    # This should never happen if we were able to successfully confirm the account using the same phone number
    if not user:
        logger.error(f"Failed to find cognito user with phone number {request.username}")
        return create_response(500, {})

    logger.info("Creating user profile")
    user_service.create_user(id=user.userId, phone_number=user.phoneNumber)
    return create_response(200, {})


@exception_handler
def forgot_password(event, context):
    body = json.loads(event.get('body', '{}'))
    request = ForgotPasswordRequest(username=body.get("username"))
    auth_service.forgot_password(request=request)
    return create_response(200, {
        "message": "A verification code has been sent to that number if it exists."
    })


@exception_handler
def confirm_forgot_password(event, context):
    body = json.loads(event.get('body', '{}'))
    request = ResetPasswordRequest(username=body.get("username"), password=body.get("password"), code=body.get("code"))
    auth_service.confirm_forgot_password(request=request)
    return create_response(200, {
        "message": "Password has been reset"
    })


@exception_handler
def send_code(event, context):
    query_params = event.get("queryStringParameters", {})
    headers = event.get('headers', {})
    access_token = headers.get("AccessToken") or headers.get("accesstoken")
    request = SendCodeRequest(access_token=access_token, attribute_name=query_params.get("attribute"))
    response = auth_service.send_code(request=request)
    return create_response(200, response)


@exception_handler
def confirm_code(event, context):
    query_params = event.get("queryStringParameters", {})
    headers = event.get('headers', {})
    access_token = headers.get("AccessToken") or headers.get("accesstoken")
    request = ConfirmCodeRequest(access_token=access_token,
                                 attribute_name=query_params.get("attribute"),
                                 code=query_params.get("code"))
    user_id = auth_service.confirm_code(request=request)
    user_service.confirm_user_email(id=user_id)
    return create_response(200, {"message": "Success"})


@exception_handler
def update_user(event, context):
    path_params = event.get('pathParameters', {})
    body = json.loads(event.get('body', '{}'))
    request = UpdateRequest(email=body.get("email"))
    updated_user = auth_service.update_user(user_id=path_params.get("userId"), request=request)
    return create_response(200, updated_user)
